### PR TITLE
Make CIO TCP half close behavior optional and disabled by default.

### PR DESCRIPTION
**Pull request by dmitrievanthony**
**Subsystem**
Client, CIO

**Motivation**
In the case of using `WriteChannelContent` CIO client closes the output channel and so triggers TCP connection half-close. Although this behavior is correct from the protocol perspective it's not encouraged in some cases and is not supported by all servers (Netty for example).

There are several issues caused by current implementation: #1456 and #1462.

**Solution**
This change makes TCP half-close optional and disabled by default.

### CODE REVIEW

File: ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/CIOEngineConfig.kt [hunk 1]
```diff
@@ -71,4 +71,11 @@ class EndpointConfig {
      * Maximum number of connection attempts.
      */
     var connectRetryAttempts: Int = 5
+
+    /**
+     * Allow socket to close output channel immediately on writing completion (TCP connection half close).
+     */
+    var allowHalfClose: Boolean = false
```
Comment (e5l) [inline ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/CIOEngineConfig.kt]: Let's turn it on by default

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/MultiPartFormDataTest.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                        "file",
+                        ByteArray(1024 * 1024),
+                        Headers.build {
+                            append(HttpHeaders.ContentDisposition, """form-data; name="file"; filename="test.png"""")
+                        }
+                    )
+                })
+            }.execute()
+
+            check(result.status == HttpStatusCode.OK)
```
Comment (e5l) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/MultiPartFormDataTest.kt]: assertEquals

File: ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/utils.kt [hunk 3]
```diff
@@ -125,3 +132,29 @@ internal suspend fun readResponse(
...
+    if (closeOnCoroutineCompletion) {
+        // Pure output represents a socket output channel that is closed when request fully processed or after
+        // request sent in case TCP half-close is allowed.
+        coroutineContext[Job]!!.invokeOnCompletion {
+            close()
+        }
+    }
+
+    return GlobalScope.reader(coroutineContext, autoFlush = true) {
+        this.channel.copyTo(this@withoutClosePropagation, Long.MAX_VALUE)
```
Comment (e5l) [inline ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/utils.kt]: this

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/MultiPartFormDataTest.kt (hunk 2 above)
Comment (dmitrievanthony) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/MultiPartFormDataTest.kt]: Done.

File: ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/utils.kt (hunk 3 above)
Comment (dmitrievanthony) [reply ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/utils.kt]: Done.

**Pull request merged**